### PR TITLE
optimize hash_tree_root using hashtree

### PR DESCRIPTION
**Pull request by arnetheduck**
brings `hash_tree_root(BeaconState)` down ~800ms to ~660ms - further optimization is possible by hashing multiple blocks in a single call, but this requires more significant surgery.

### CODE REVIEW

File: ssz_serialization/merkleization.nim [hunk 1]
```diff
@@ -120,23 +120,82 @@ func digest*(a: openArray[byte]): Digest {.noinit.} =
...
+type Hasher = proc(output: pointer, input: pointer, count: uint64) {.cdecl, noSideEffect, raises: [Defect].}
+
+func digest64x(output: pointer, input: pointer, count: uint64) {.cdecl, noSideEffect, raises: [Defect].} =
+  cast[ptr Digest](output)[] =
+     digest(cast[ptr UncheckedArray[byte]](input).toOpenArray(0, 63))
+  discard
+
+func best_sha256_implementation(): Hasher {.importc.}
+
+let digest64* = block:
```
Comment (mratsim) [inline ssz_serialization/merkleization.nim:147]: digest64 is a really bad name. Maybe `digest2x` or `digest_2way`?

File: ssz_serialization/merkleization.nim [hunk 2]
```diff
@@ -120,23 +120,82 @@ func digest*(a: openArray[byte]): Digest {.noinit.} =
...
+    else:
+      result = computeDigest:
+        h.update a
+        h.update b
+  trs "HASH RESULT ", result
 
-    h.update a
-    h.update b
-    h.update c
+func digest(a, b, c: openArray[byte]): Digest {.noinit.} =
```
Comment (mratsim) [inline ssz_serialization/merkleization.nim:175]: When does this happen?

Comment (arnetheduck) [reply ssz_serialization/merkleization.nim:175]: when joining a 32-byte data chunk with a <32-byte chunk - `c` is the zero padding mandated by ssz

File: ssz_serialization/merkleization.nim (hunk 1 above)
Comment (arnetheduck) [reply ssz_serialization/merkleization.nim:147]: well .. it's 64 bytes which is the significant requirement here .. ie unlike the other digest functions, it only works for blocks of that length.

Comment (cheatfate) [reply ssz_serialization/merkleization.nim:147]: just call it `digestBlockSize` :)

**Pull request merged**